doc/man/rnexplorer.man
.TH Rotifer 0 rnexplorer
.SH NAME
.PP
rnexplorer \- Annotate, clean, reformat rneighbors output
.SH SYNOPSIS
.PP
\fBrnexplorer\fP [\fIOPTIONS\fP] [\fIFILE\fP]
.PP
# Simplest usage:
.PP
\fBrnexplorer\fP rneighbors.tsv \-of gi2operon > rneighbors.gi2operon
.SH DESCRIPTION
.PP
This program helps you explore snapshots of gene neighborhooods created by
.br
rneighbors/acc2operon (Rotifer) and/or gi2operons (TASS)
.SH AUTHORS
.PP
Gilberto Kaihami; Robson Souza
.SH OPTIONS
.PP
This section describes program options, aliases and default values
.PP
\fBProgram options\fP
.PP
\fBfile\fP
.PP
Input table/gi2operon file. Pipe compatible
.PP
default: empty
.PP
\fB\-\-header\fP, \fB\-y\fP
.PP
Add more information to gi2operon header. The additional information must be a valid collumn
.br
in the table format
.PP
example: rnexplorer \-y assembly \-y classification
.PP
default: empty
.PP
\fB\-\-outformat\fP, \fB\-of\fP
.PP
Set output format. Acceptable output formats are:
.PP
.RS
.nf
\- table:     all neighborhood data formatted as a text table
\- gi2operon: all neighborhood data formatted as gi2operon
\- compact:   all neighborhood data formatted as compact
.fi
.RE
.PP
default: same as input
.PP
\fB\-\-exclude_by_type\fP, \fB\-xt\fP
.PP
Exclude individual features type in the format KEY:VALUE.
.br
The value could be an unique value (e.g. \-xt type:CDS) or a file containing one column
.br
with values in each line.
.PP
example: \-xt type:CDS
         \-xt seq_type:[FILE]
         \-xt seq_type:plasmid
.PP
\fB\-\-include_by_type\fP, \fB\-it\fP
Include individual features type in the format KEY:VALUE.
.br
The value could be an unique value (e.g. \-it type:CDS) or a file containing one column
.br
with values in each line.
.PP
example: \-it type:CDS
         \-it seq_type:[FILE]
         \-it seq_type:plasmid
.PP
\fB\-\-annotation\fP, \fB\-a\fP
.PP
Add annotation column to the output.  
.PP
Usage formats: [FILE] or [COL_KEY], [COL_KEY:FILE], or [COL_KEY:FILE:COL_NAME]
.br
The annotation file must contain at least two columns. The first column is the key and
.br
the second colum contains the annotation.
.br
There are three ways to use this flag:
.br
(1) Input just the annotation file it will map the first column against the pid column
.br
in the neighborhood table.
.br
(2) Include a column in the output table. The column can be selected by number, column
.br
name or interval (see exclude_column)
.br
(3) You can pass the column to map and the file containing the annotation.
.br
(4) You can pass 3 parameters, the first is the reference column in the neighborhood
.br
table, the second is the annotation file, and the last parameter is the new column name.  
.PP
examples:
.PP
.RS
.nf
(1) \-a [FILE]    
(2) \-a pfam    
(2) \-a 0..3    
(2) \-a 1    
(3) \-a pid:[FILE]    
(3) \-a nucleotide:[FILE]    
(4) \-a pid:[FILE]:annotation\\_col    
.fi
.RE
.PP
default: empty
.PP
\fB\-\-padtable\fP, \fB\-p\fP
.PP
Format gi2operon output
.PP
default: empty
.PP
\fB\-\-exclude_column\fP, \fB\-xc\fP
.PP
Exclude a column in the output table
.br
Select a column by number (index starts from 0), column name or an interval
.br
Usage formats: col_number, start..end, column name  
.PP
examples: \-xc 0       (refers to the first column)
          \-xc 0..4    (refers from the first to the fifth column)
          \-xc product (refers to product column)
.PP
\fB\-\-include_column\fP, \fB\-ic\fP
.PP
Include a column in the output table
.br
Select a columb by number (index starts from 0), column name or an interval
.br
Usage formats: col_number, start..end, column name  
.PP
examples: \-ic 0       (refers to the first column)
          \-ic 0..4    (refers from the first to the fifth column)
          \-ic product (refers to product column)
.PP
\fB\-\-include_column\fP, \fB\-ic\fP
.PP
Include a column in the output table
.br
Select a columb by number (index starts from 0), column name or an interval
.br
Usage formats: col_number, start..end, column name  
.PP
examples: \-ic 0       (refers to the first column)
          \-ic 0..4    (refers from the first to the fifth column)
          \-ic product (refers to product column)
.PP
\fB\-\-optionargs\fP, \fB\-oa\fP
.PP
Advanced options.
.br
Select a architecture column for compact format in the formart 'arch':[COLUMN_NAME] or 'arch'=[COLUMN_NAME]  
.PP
example: rnexplorer \-of compact \-oa arch:pfam <file>
.br
         rnexplorer \-of compact \-oa arch=pfam \-a pid:arch.tsv:pfam <file>  
.SH Program options summary
.TS
allbox;
cb cb cb
r r c
r r c
r r c
r r c
r r c
r r c
r r c
r r c
r r c
r r c
.
Long name	Aliases	Type
\-\-header	\-y	list
\-\-outformat	\-of	string
\-\-exclude_by_type	\-xt	list
\-\-include_by_type	\-it	list
\-\-annotation	\-a	list
\-\-padtable	\-p	boolean
\-\-exclude_column	\-xc	list
\-\-include_column	\-it	list
\-\-optionargs	\-oa	hash
\-\-version		boolean
.TE


dev/old/rnexplorer.0d02.py
#!/usr/bin/env python3

import pandas as pd
import sys
import os
sys.path.insert(0, os.path.join('/home/kaihami/mymodules'))
import rotifer.core.cli as corecli
import argparse
from multiprocessing import Process
import warnings
import numpy as np
warnings.filterwarnings('ignore')

__version__ = 0.02
__authors__ = 'Gilberto Kaihami'

def parse_cli():
    parser = argparse.ArgumentParser(description = 'Convert rneighbors output (table<->gi2operons)',
                                     formatter_class = argparse.RawTextHelpFormatter)
    parser.add_argument('file',
                        help = 'Input neighborhood file',
                        action = corecli.action.autoload,
                        nargs = '*',
                        duplicates = False)

    parser.add_argument('-y', '--header',
                              action='append',
                              help = 'Add more information to gi2operon header. The additional information must be a valid collumn in the table format\n\
(example: -y assembly, -y classification -y seq_type)',
                              default = [])

    parser.add_argument('-of', '--outformat',
                        dest = 'outformat',
                        help = 'Output format (table, gi2ioperon, not implemented)',
                        default = '')

    parser.add_argument('--version',
                        action = 'version',
                        version = corecli.version(program = 'rnexplorer',
                                                  version = __version__,
                                                  authors = __authors__,
                                                  description = 'Convert neighborhood output (table<->gi2operons)'
                                                  ))
    parser.add_argument('--exclude_by_type', '--xt',
                        '-exclude_by_type', '-xt',
                        dest = 'excludebytype',
                        default = [],
                        help = 'Exclude individual features by feature type. (e.g. tRNA)',
                        action = 'append',
                        type = str)

    args = parser.parse_args()

    return args

def collen(df):
    df['start'].astype(str)
    df['end'].astype(str)
    df['plen'].astype(str)
    q = '-->'
    df['cds_loc'] = df['start'].astype(str)+'..'+df['end'].astype(str)
    direction = len('dir')
    len_plen = df['plen'].astype(str).str.len().max()
    len_pid = df['pid'].str.len().max() if df['pid'].str.len().max() >= len('pid') else len('pid')
    len_type = df['type'].str.len().max() if df['type'].str.len().max() >= len('type') else len('type')
    len_gene = df['gene'].str.len().max() if df['gene'].str.len().max() >= len('gene') else len('gene')
    len_cds = df['cds_loc'].str.len().max()
    len_locus = df['locus'].str.len().max() if df['locus'].str.len().max() >= len('locus') else len('locus')
    try:
        len_modified = df['modified'].str.len().max() if df['modified'].str.len().max() >= len('gi') else len('gi')
    except:
        len_modified = 1
    len_product = 0
    collen_len = [len(q), len_cds, len('dir'), len_plen, len_pid, len_type,
                  len_gene, len_locus, len_modified, len_product]

    return collen_len

def gi2operon(sub_df, other_info = []):

    sub_df['plen'].astype(str)
    if '' in sub_df['plen'].values:
        sub_df.loc[sub_df[sub_df['plen'] == ''].index, 'plen'] = '.'

    if '' in sub_df['pid']:
        sub_df.loc[sub_df[sub_df['pid'] == ''].index, 'pid'] = '.'

    col_len = collen(sub_df)
    protein = sub_df[sub_df['query'] == '1']['pid'].values[0]
    nucleotide = sub_df.nucleotide.values[0]

    max_header = 'ORGANISM ' + df.organism.values[0] + ' accession no is ' + nucleotide + ' Protein is '+ protein
    if other_info:
        for e in other_info:
            max_header += ' | '+ e + ':'+''.join(str(sub_df[e].values[0]))

    print(max_header)

    header = [".","cds","dir","len","pid","type","gene","locus","gi",'product']
    header = [header[int(i)].ljust(int(col_len[int(i)])) for i in range(0,len(header), 1)]
    print('  '.join(header))

    # sub_df['.'] = np.where(sub_df['query'] == 1, '-->', '.')
    # sub_df['dir'] = np.where(sub_df['strand'] == 1, '+', '-')
    # sub_df['cds'] = sub_df['start'].astype(str) + '..' + sub_df['end'].astype(str)
    # sub_df['len'] =sub_df['plen']
    # sub_df['gi'] = sub_df.shape[0]*['.']
    # print(sub_df.to_string(columns = header, index = False, justify = 'left'))
    #
    for i, row in sub_df.iterrows():
        query = '-->' if row.pid in sub_df[sub_df['query'] == 1].pid.values else '.'
        direction = '+' if row.strand == '1' else '-'

        cds = str(row.start) +'..'+str(row.end)
        toprint = [str(x) for x in [query, cds, direction,
                          row.plen, row.pid, row.type,
                          row.gene, row.locus, '.', row['product']]]
        toprint = [toprint[i].ljust(int(col_len[i])) for i in range(len(toprint))]
        print('  '.join(toprint))
    print("---------------------------------------")

def table(f):
    ipt = f

    block_id = 0
    header = 'nucleotide start end strand block_id, query pid type plen locus seq_type assembly gene product organism classification'.split()
    print('\t'.join(header))

    for line in ipt:
#        print(line)
        if 'ORGANISM' in line:
            block_id +=1
            chr_type = '.'
            asm = '.'
            classification = '.'
            organism = line.split('ORGANISM')[1].split(' accession no')[0].strip()
            nucleotide = line.split('ORGANISM')[1].split(' accession no is')[1].split('Protein')[0].strip()
            if '|' in line:
                other_information = line.split('|')
                for ele in other_information:
                    if 'seq_type' in ele:
                        chr_type = ele.split(':')[1].strip()
                    if 'assembly' in ele:
                        asm = ele.split(':')[1].strip()
                    if 'classification' in ele:
                        classification = ele.split(':')[1].strip()

        if 'ORGANISM' not in line and 'cds' not in line and '---------------------' not in line:
            line = [x for x in line.split(' ') if x != '']
#            print(line)
            query, start_end, strand, length, pid,cds_type, gene, locus, gi, *product = line
#            print (line)
            start, end = start_end.split('..')
            query = 1 if query == '-->' else 0
            strand = 1 if strand == '+' else -1

            print('\t'.join([nucleotide, start,end, str(strand), str(block_id), str(query),
                             pid, cds_type, length, locus, chr_type, asm, gene,
                             ' '.join(product), organism, classification]))


def filter_by(df, args):
    df[~(df['type'].isin(args))]
    return df

if __name__ == '__main__':
    args = parse_cli()
    # to_open = open(args.file)
    to_open = args.file

    other_info = args.header


    if 'ORGANISM' in to_open[0]:
        input_format_type = 'gi2operon'

    else:
        input_format_type = 'table'
    if args.outformat:
        outformat = args.outformat
    else:
        outformat = 'table' if input_format_type == 'gi2operon' else 'gi2operon' 
    if input_format_type == 'table':
        # print(to_open)
        s = [x.split('\t') for x in to_open]

        df = pd.DataFrame(s[1:], columns = s[0])

        df = df.fillna('.')
    
        g = df.groupby('block_id')
        for i, group in g:
            group = group.fillna('.')
            gi2operon(group, other_info) # print gi2operon format

    if input_format_type == 'gi2operon':
#        print(to_open)
        table(to_open) # print table format

